miasm2/core/asmblock.py

```diff
@@ -359,132 +323,209 @@ class AsmSymbolPool(object):
-        if (label.name in self._name2label and
-                label != self._name2label[label.name]):
-            raise ValueError('symbol %s has same name as %s' %
-                             (label, self._name2label[label.name]))
-
-        self._labels.add(label)
-        if label.offset is not None:
-            self._offset2label[label.offset] = label
-        if label.name != "":
-            self._name2label[label.name] = label
-        return label
-
-    def remove_label(self, label):
+        known_loc_key = self.getby_name(name)
+        if known_loc_key is not None:
+            known_offset = self.loc_key_to_offset(known_loc_key)
+            if known_offset != offset:
+                raise ValueError(
+                    'symbol %s with different offset %s %s' % (
+                        name, offset, known_offset
+                    )
+                )
+            return known_loc_key
+
+        if offset is not None:
+            known_loc_key = self.getby_offset(offset)
+            if known_loc_key:
+                known_name = self.loc_key_to_name(known_loc_key)
+                if known_name != name:
+                    raise ValueError(
+                        'offset %s with different names %s %s' % (
+                            offset, name, known_name
+                        )
+                    )
+                return known_loc_key
+
+        loc_key = LocKey(self._loc_key_num)
+        self._loc_key_num += 1
+
+        self._loc_keys.add(loc_key)
+
+        if offset is not None:
+            assert offset not in self._offset_to_loc_key
+            self._offset_to_loc_key[offset] = loc_key
+            self._loc_key_to_offset[loc_key] = offset
+
+        self._name_to_loc_key[name] = loc_key
+        self._loc_key_to_name[loc_key] = name
+        return loc_key
+
+    def remove_location(self, loc_key):
         """
-        Delete a @label
+        Delete a @loc_key
         """
-        self._name2label.pop(label.name, None)
-        self._offset2label.pop(label.offset, None)
-        if label in self._labels:
-            self._labels.remove(label)
+        name = self._loc_key_to_name.pop(loc_key, None)
+        _ = self._name_to_loc_key.pop(name, None)
+
+        offset = self._loc_key_to_offset.pop(loc_key, None)
+        _ = self._offset_to_loc_key.pop(offset, None)
+
+        self._loc_keys.remove(loc_key)
 
-    def del_label_offset(self, label):
-        """Unpin the @label from its offset"""
-        self._offset2label.pop(label.offset, None)
-        label.offset = None
+    def del_loc_key_offset(self, loc_key):
+        """Unpin the @loc_key from its offset"""
+        offset = self._loc_keys_to_offset.pop(loc_key)
+        _ = self._offset_to_loc_key.pop(offset, None)
 
     def getby_offset(self, offset):
-        """Retrieve label using its @offset"""
-        return self._offset2label.get(offset, None)
+        """Retrieve loc_key using its @offset"""
```

Comment:
Like the comment above, it would be great the reflect the `None` behavior in the docstring. Indeed, an user might exepect a KeyError to be raised.
In addition, it could be present in the main class doc, as a list on how a `loc_key` can be retrivied

Reply:
done